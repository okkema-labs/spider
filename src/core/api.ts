import { fromHono, type RouterOptions } from "chanfana"
import { Hono } from "hono"
import {
  authenticate,
  error,
  type AuthBindings,
  type AuthVariables,
} from "../middleware"

type APIInit = {
  tokenUrl: string
  options?: RouterOptions
  scopes?: Record<string, string>
}

type APIBindings = AuthBindings

type APIVariables = AuthVariables

export function API<
  Bindings extends APIBindings,
  Variables extends APIVariables,
>({ tokenUrl, options, scopes }: APIInit) {
  const base = new Hono<{ Bindings: Bindings; Variables: Variables }>()
  const app = fromHono(base, options)

  app.registry.registerComponent("securitySchemes", "Oauth2", {
    type: "oauth2",
    flows: {
      clientCredentials: {
        tokenUrl,
        scopes,
      },
    },
  })

  app.use("*", authenticate)
  app.onError(error)
  return app as Hono<{ Bindings: Bindings; Variables: Variables }>
}
