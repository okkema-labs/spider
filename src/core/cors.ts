type CORS = {
  origin: string
  methods?: string[]
  headers?: string[]
}

const DEFAULT: CORS = {
  origin: "*",
  methods: ["*"],
  headers: ["*"],
}

const CORS = (
  request: Request,
  response: Response,
  cors: Partial<CORS>,
): Response => {
  const { method } = request
  const { origin, methods, headers } = { ...DEFAULT, ...cors }
  const clone = new Response(response.body, response)
  clone.headers.append("Access-Control-Allow-Origin", origin)
  if (method === "OPTIONS") {
    if (methods.length)
      response.headers.append(
        "Access-Control-Allow-Methods",
        methods.join(", "),
      )
    if (headers.length)
      response.headers.append(
        "Access-Control-Allow-Headers",
        headers.join(", "),
      )
  }
  return clone
}

export default CORS
