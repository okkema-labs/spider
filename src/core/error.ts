type CoreErrorInit = Partial<Omit<CoreError, "name" | "message" | "stack">>

class CoreError extends Error {
  type: string
  title: string
  status: number
  detail: string

  constructor(init?: CoreErrorInit) {
    super()
    // Error
    this.name = this.constructor.name
    this.message = init?.title ?? this.constructor.name
    // Problem Details
    this.type = init?.type ?? this.constructor.name
    this.title = init?.title ?? this.constructor.name
    this.status = init?.status ?? 500
    this.detail = init?.detail ?? this.constructor.name
  }
}

export default CoreError
