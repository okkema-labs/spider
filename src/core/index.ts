import Worker, {
  FetchEventHandler,
  ScheduledEventHandler,
  Logger,
} from "./worker"
import Router, { RequestHandler, Route } from "./router"
import Problem from "./problem"
export { Worker, Router, Problem }
export type {
  FetchEventHandler,
  ScheduledEventHandler,
  Logger,
  RequestHandler,
  Route,
}
export default Worker
