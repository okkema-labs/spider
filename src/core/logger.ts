/* eslint-disable @typescript-eslint/no-explicit-any */
import { captureMessage, type SeverityLevel } from "@sentry/cloudflare"

type Logger = {
  debug(...args: any[]): void
  error(...args: any[]): void
  info(...args: any[]): void
  warn(...args: any[]): void
}

function sendToSentry(level: SeverityLevel, ...args: any[]) {
  captureMessage(
    `${args.reduce(function (prev, curr) {
      if (typeof curr === "object") prev += ` ${JSON.stringify(curr)}`
      else prev += ` ${curr}`
      return prev
    }, "")}`,
    level,
  )
}

export const Logger: Logger = {
  debug(...args) {
    console.debug(...args)
    sendToSentry("debug", ...args)
  },
  error(...args) {
    console.error(...args)
    sendToSentry("error", ...args)
  },
  info(...args) {
    console.info(...args)
    sendToSentry("info", ...args)
  },
  warn(...args) {
    console.warn(...args)
    sendToSentry("warning", ...args)
  },
}
