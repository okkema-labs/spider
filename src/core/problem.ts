type ProblemInit = Pick<Problem, "type" | "title" | "status" | "detail">

class Problem extends Error {
  title: string
  detail: string
  type?: string
  status?: number

  constructor(init: ProblemInit) {
    super()
    // Error
    this.name = "Problem"
    this.message = init.title
    // Problem Details
    this.type = init.type
    this.title = init.title
    this.status = init.status ?? 500
    this.detail = init.detail
  }
}

export default Problem
