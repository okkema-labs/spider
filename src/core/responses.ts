export const Ok = (body?: BodyInit): Response =>
  new Response(body, { status: 200, statusText: 'Ok' })
