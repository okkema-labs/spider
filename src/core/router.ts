import Problem from "./problem"

export type RoutedRequest = Request & {
  query?: { [key: string]: string }
  params?: { [key: string]: string }
}

type RequestHandler = (request: RoutedRequest) => Promise<Response>

type Router = {
  routes: Route[]
  handle: RequestHandler
  use: (route: Route) => void
  get: (path: string, ...handlers: RequestHandler[]) => void
  post: (path: string, ...handlers: RequestHandler[]) => void
  put: (path: string, ...handlers: RequestHandler[]) => void
  del: (path: string, ...handlers: RequestHandler[]) => void
  all: (path: string, ...handlers: RequestHandler[]) => void
}

type HttpMethod = "GET" | "POST" | "PUT" | "DELETE" | "ALL"

export type Route = {
  method: HttpMethod
  path: string
  handlers: RequestHandler[]
  regex?: RegExp
}

const Router = (): Router => {
  const routes: Route[] = []

  const use = (route: Route) => {
    routes.push({
      regex:
        route.regex ??
        RegExp(
          `^${route.path
            .replace(/(\/?)\*/g, "($1.*)?")
            .replace(/\/$/, "")
            .replace(/:(\w+)(\?)?(\.)?/g, "$2(?<$1>[^/]+)$2$3")
            .replace(/\.(?=[\w(])/, "\\.")}/*$`,
        ),
      ...route,
    })
  }

  const get = (path: string, ...handlers: RequestHandler[]) =>
    use({
      method: "GET",
      path,
      handlers,
    })

  const post = (path: string, ...handlers: RequestHandler[]) =>
    use({
      method: "POST",
      path,
      handlers,
    })

  const put = (path: string, ...handlers: RequestHandler[]) =>
    use({
      method: "PUT",
      path,
      handlers,
    })

  const del = (path: string, ...handlers: RequestHandler[]) =>
    use({
      method: "DELETE",
      path,
      handlers,
    })

  const all = (path: string, ...handlers: RequestHandler[]) =>
    use({
      method: "ALL",
      path,
      handlers,
    })

  const handle = async (request: RoutedRequest) => {
    const url = new URL(request.url)
    request.query = Object.fromEntries(url.searchParams)
    for (const { method, regex, handlers } of routes) {
      const match = url.pathname.match(regex)
      if (match && (method === request.method || method === "ALL")) {
        request.params = match.groups
        for (const handler of handlers) {
          const response = await handler(request)
          if (response !== undefined) return response
        }
        throw new Problem({
          detail: "Please contact the admin with the request details.",
          status: 500,
          title: "The route handlers did not return a response",
          type: "NoResponse",
        })
      }
    }
  }

  return {
    routes,
    handle,
    use,
    get,
    post,
    put,
    del,
    all,
  }
}

export default Router
