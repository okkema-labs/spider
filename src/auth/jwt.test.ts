import JWT from "./jwt"
import { Problem } from "../core"

jest.mock("./jwt", () => {
  const JWT = jest.requireActual("./jwt").default
  return { ...JWT, decode: jest.fn(JWT.decode) }
})

const createJWT = () => ({
  header: {
    alg: "RS256",
    kid: "",
    typ: "JWT",
  },
  payload: {
    aud: "audience",
    exp: Date.now() + 1000,
    iss: "issuer",
    sub: "subject",
  },
  signature: "",
})

describe("JWT", () => {
  describe("decode", () => {
    it("decodes a valid token", () => {
      const token =
        "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c"
      const result = JWT.decode(token)
      expect(result).toBeDefined()
      expect(result).toMatchSnapshot()
    })
    it("throws an error when token is invalid", () => {
      const token = "some.bad.token"
      expect(() => JWT.decode(token)).toThrow(Problem)
    })
  })

  describe("validate", () => {
    it("throws an error if the type is invalid", async () => {
      const token = "token"
      const audience = "audience"
      const issuer = "issuer"
      // const jwt = createJWT()
      // jwt.header.typ = "invalid"
      // const mockDecode = JWT.decode as jest.MockedFunction<typeof JWT.decode>
      // mockDecode.mockImplementationOnce((token) => jwt)
      await expect(JWT.validate(token, audience, issuer)).rejects.toThrow(
        Problem,
      )
    })
    it("throws an error if the algorithm is invalid", async () => {
      const token = "token"
      const audience = "audience"
      const issuer = "issuer"
      const jwt = createJWT()
      jwt.header.alg = "invalid"
      const mockDecode = JWT.decode as jest.MockedFunction<typeof JWT.decode>
      mockDecode.mockImplementationOnce((token) => jwt)
      await expect(JWT.validate(token, audience, issuer)).rejects.toThrow(
        Problem,
      )
    })
    it("throws an error if the key id is missing", async () => {
      const token = "token"
      const audience = "audience"
      const issuer = "issuer"
      const jwt = createJWT()
      jwt.header.kid = ""
      const mockDecode = JWT.decode as jest.MockedFunction<typeof JWT.decode>
      mockDecode.mockImplementationOnce((token) => jwt)
      await expect(JWT.validate(token, audience, issuer)).rejects.toThrow(
        Problem,
      )
    })
    it("throws an error if the audience is invalid", async () => {
      const token = "token"
      const audience = "audience"
      const issuer = "issuer"
      const jwt = createJWT()
      jwt.payload.aud = "invalid"
      const mockDecode = JWT.decode as jest.MockedFunction<typeof JWT.decode>
      mockDecode.mockImplementationOnce((token) => jwt)
      await expect(JWT.validate(token, audience, issuer)).rejects.toThrow(
        Problem,
      )
    })
    it("throws an error if the issuer is invalid", async () => {
      const token = "token"
      const audience = "audience"
      const issuer = "issuer"
      const jwt = createJWT()
      jwt.payload.iss = "invalid"
      const mockDecode = JWT.decode as jest.MockedFunction<typeof JWT.decode>
      mockDecode.mockImplementationOnce((token) => jwt)
      await expect(JWT.validate(token, audience, issuer)).rejects.toThrow(
        Problem,
      )
    })
    it("throws an error if the expiration is missing", async () => {
      const token = "token"
      const audience = "audience"
      const issuer = "issuer"
      const jwt = createJWT()
      jwt.payload.exp = 0
      const mockDecode = JWT.decode as jest.MockedFunction<typeof JWT.decode>
      mockDecode.mockImplementationOnce((token) => jwt)
      await expect(JWT.validate(token, audience, issuer)).rejects.toThrow(
        Problem,
      )
    })
    it("throws an error if the jwt is expired", async () => {
      const token = "token"
      const audience = "audience"
      const issuer = "issuer"
      const jwt = createJWT()
      jwt.payload.exp = Date.now()
      const mockDecode = JWT.decode as jest.MockedFunction<typeof JWT.decode>
      mockDecode.mockImplementationOnce((token) => jwt)
      await expect(JWT.validate(token, audience, issuer)).rejects.toThrow(
        Problem,
      )
    })
    it.skip("throws an error if no matching key is found", async () => {
      // const keyStore = JWK.createKeyStore()
      // await keyStore.generate("RSA", 2048, { alg: "RS256", use: "sig" })
      // const token = "token"
      // const audience = "audience"
      // const issuer = "issuer"
      // const jwks = keyStore.toJSON() as { keys: JWK[] }
      // const jwt = createJWT()
      // const mockDecode = decode as jest.MockedFunction<typeof decode>
      // mockDecode.mockImplementationOnce((token) => jwt)
      // const mockFetchJWKS = fetchJWKS as jest.MockedFunction<typeof fetchJWKS>
      // mockFetchJWKS.mockImplementationOnce((issuer) => Promise.resolve(jwks))
      // await expect(validateToken(token, audience, issuer)).rejects.toThrow(
      //   Problem,
      // )
      // expect(decode).toHaveBeenCalledWith(token)
    })
    // it("throws an error if the signature is invalid", async () => {
    //   const keyStore = JWK.createKeyStore()
    //   await keyStore.generate("RSA", 2048, { alg: "RS256", use: "sig" })
    //   const token = "token"
    //   const audience = "audience"
    //   const issuer = "issuer"
    //   const jwks = keyStore.toJSON() as {keys: JWK[]}
    //   const jwt = createJWT()
    //   jwt.header.kid = jwks.keys[0].kid
    //   const mockDecode = decode as jest.MockedFunction<typeof decode>
    //   mockDecode.mockImplementationOnce((token) => jwt)
    //   const mockFetchJWKS = fetchJWKS as jest.MockedFunction<typeof fetchJWKS>
    //   mockFetchJWKS.mockImplementationOnce((issuer) =>
    //     Promise.resolve(jwks),
    //   )
    //   await expect(validateToken(token, audience, issuer)).rejects.toThrow(
    //     Problem,
    //   )
    //   expect(decode).toHaveBeenCalledWith(token)
    // })
    it.skip("returns true if the token is vaild", async () => {
      // const keyStore = JWK.createKeyStore()
      // const key = await crypto.subtle.generateKey(
      //   {
      //     name: "RSASSA-PKCS1-v1_5",
      //     hash: "SHA-256",
      //     publicExponent: new Uint8Array([0x01, 0x00, 0x01]),
      //     modulusLength: 2048,
      //   } as any,
      //   true,
      //   ["sign", "verify"],
      // )
      // const jwk = await crypto.subtle.exportKey("jwk", key.publicKey)
      // const audience = "audience"
      // const issuer = "issuer"
      // const header = base64.encode(
      //   JSON.stringify({ alg: "RS256", typ: "JWT", kid: "123" }),
      // )
      // const payload = base64.encode(
      //   JSON.stringify({
      //     aud: audience,
      //     iss: issuer,
      //     exp: Date.now() + 1000,
      //   }),
      // )
      //  uint8 to string
      // console.log(`${header.replace("=", "")}.${payload.replace("=", "")}`)
      // const token = await crypto.subtle.sign(
      //   { name: "RSASSA-PKCS1-v1_5" },
      //   key.privateKey,
      //   new TextEncoder().encode(`${header}.${payload}`),
      // )
      // console.log(token)
      // const signature = new TextDecoder().decode(token)
      // console.log(signature)
      // const token = await JWS.createSign(
      //   {
      //     alg: "RS256",
      //     format: "compact",
      //     fields: {
      //       typ: "JWT",
      //       kid: key.kid,
      //     },
      //   },
      //   key,
      // )
      //   .update(
      //     JSON.stringify({
      //       aud: audience,
      //       iss: issuer,
      //       exp: Date.now() + 1000,
      //     }),
      //   )
      //   .final()
      //   .then((result) => String(result))
      // const jwks = keyStore.toJSON() as { keys: JWK[] }
      // const mockDecode = decode as jest.MockedFunction<typeof decode>
      // mockDecode.mockImplementationOnce((token) => {
      //   // TODO find better way of using unmocked implementation
      //   const [header, payload, signature] = token.split(".")
      //   return {
      //     header: JSON.parse(atob(header)),
      //     payload: JSON.parse(atob(payload)),
      //     signature: btoa(
      //       encodeURIComponent(signature).replace(
      //         /%([0-9A-F]{2})/g,
      //         (match, p1) => {
      //           return String.fromCharCode(parseInt(p1, 16))
      //         },
      //       ),
      //     ),
      //   }
      // })
      // const mockFetchJWKS = fetchJWKS as jest.MockedFunction<typeof fetchJWKS>
      // mockFetchJWKS.mockImplementationOnce((issuer) => Promise.resolve(jwks))
      // const result = await validateToken(token, audience, issuer)
      // expect(result).toBe(true)
      // expect(decode).toHaveBeenCalledWith(token)
    })
  })
})
