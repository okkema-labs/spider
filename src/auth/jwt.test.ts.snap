// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JWT decode decodes a valid token 1`] = `
Object {
  "header": Object {
    "alg": "HS256",
    "typ": "JWT",
  },
  "payload": Object {
    "iat": 1516239022,
    "name": "John Doe",
    "sub": "1234567890",
  },
  "signature": "U2ZsS3h3UkpTTWVLS0YyUVQ0ZndwTWVKZjM2UE9rNnlKVl9hZFFzc3c1Yw==",
}
`;
